lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,392 @@
+use crate::TxSyncError;
+
+use lightning::util::logger::Logger;
+use lightning::{log_error, log_given_level, log_info, log_internal, log_debug, log_trace};
+use lightning::chain::WatchedOutput;
+use lightning::chain::{Confirm, Filter};
+
+use bitcoin::{BlockHash, BlockHeader, Script, Transaction, Txid};
+
+use esplora_client::Builder;
+#[cfg(feature = "async-interface")]
+use esplora_client::r#async::AsyncClient;
+#[cfg(not(feature = "async-interface"))]
+use esplora_client::blocking::BlockingClient;
+
+use std::sync::Mutex;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::collections::HashSet;
+use core::ops::Deref;
+
+/// Synchronizes LDK with a given [`Esplora`] server.
+///
+/// Needs to be registerd with a [`ChainMonitor`] via the [`Filter`] interface to be informed of
+/// transactions and outputs to montor.
+///
+/// [`Esplora`]: https://github.com/Blockstream/electrs
+/// [`ChainMonitor`]: lightning::chain::chainmonitor::ChainMonitor
+/// [`Filter`]: lightning::chain::Filter
+pub struct EsploraSyncClient<L: Deref>
+where
+	L::Target: Logger,
+{
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	queued_transactions: Mutex<HashSet<Txid>>,
+	// Transactions that were previously processed, but must not be forgotten yet.
+	watched_transactions: Mutex<HashSet<Txid>>,
+	// Outputs that were registered via the `Filter` interface and have to be processed.
+	queued_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Outputs that were previously processed, but must not be forgotten yet.
+	watched_outputs: Mutex<HashSet<WatchedOutput>>,
+	// Indicates whether we need to resync, e.g., after encountering an error.
+	pending_sync: AtomicBool,
+	// The tip hash observed during our last sync.
+	#[cfg(not(feature = "async-interface"))]
+	last_sync_hash: std::sync::Mutex<Option<BlockHash>>,
+	#[cfg(feature = "async-interface")]
+	last_sync_hash: futures::lock::Mutex<Option<BlockHash>>,
+	#[cfg(not(feature = "async-interface"))]
+	client: BlockingClient,
+	#[cfg(feature = "async-interface")]
+	client: AsyncClient,
```

Comment:
Would you mind adding a `.client()` method that returns a reference to the contained `&AsyncClient` or `&BlockingClient`? This way, users who want to call other esplora methods unrelated to LDK tx sync can share the underlying (e.g. `reqwest`) connection pool.

`esplora-client` exposes a similar method to get the `reqwest` `Client` or `ureq` `Agent` for `async` / `blocking` respectively.

Reply:
Ah, so BDK offers a [`from_client`](https://docs.rs/bdk/latest/bdk/blockchain/struct.EsploraBlockchain.html#method.from_client) constructor which wraps an existing `AsyncClient` / `BlockingClient`. So the connection pool can be shared between BDK + LDK as well.

Reply:
I don't mind at all, in fact this was already on my bucket list. Now added both with https://github.com/lightningdevkit/rust-lightning/pull/1870/commits/e63d16ce1dd060b1a1767e7766137c95799e0fed, let me know how it works for you.